### PR TITLE
Instant reaction to changing options

### PR DESCRIPTION
**Pull request by ximeric**

### CODE REVIEW

**Comment by Drag13** [comment]
Hi! Thanks for the PR, I will take a look at it in a next few days. 
Btw, seems it is not passing EsLint validation. You can double check this using `npm run lint` command.

**Comment by ximeric** [comment]
As for another eslint error - I suggest change const declarations to mode "one const declaration per block", because with this mode all constants will be located in one place and no need to search theirs assignment (I can prepare new patch with PR for this).

File: src/sanitizer.js [hunk 1]
```diff
@@ -232,7 +233,76 @@
...
+    }
+
+    /**
+     * Subscribe to settings change event, watched by settings key
+     * @param {string} key
+     * @param {Function} handler
+     */
+    function onSettingsChange(key, handler) {
+        chrome.storage.onChanged.addListener((changes) => {
+            const nv = changes.settings.newValue[key], ov = changes.settings.oldValue[key];
```
Comment (Drag13) [inline src/sanitizer.js]: This code produces null exception if oldValue is null. This case will be valid for all new users.

File: src/sanitizer.js [hunk 2]
```diff
@@ -232,7 +233,76 @@
...
+        });
+    }
+
+    /**
+     * Reaction on switching isQuickActionsOn on options page
+     * @param {object} newValue New value of setting
+     */
+    function onQuickActionsVisibilityChange(newValue) {
+        if ( newValue )
+            visibleArticles.forEach(addQuickActionButtons);
```
Comment (Drag13) [inline src/sanitizer.js]: This code uses global variable `visibleArticles` which complicates code maintaining. Using non-read-only global variables should be limited to the absolutely necessary cases only

File: src/sanitizer.js [hunk 3]
```diff
@@ -232,7 +233,76 @@
...
+            (article) =>
+                belongsToAuthor(article, searchTerm) ||
+                belongsToBlog(article, searchTerm) ||
+                belongsToHab(article, searchTerm)
+        );
+
+        log(`Found ${articlesToBeRestored.length} articles to be restored for '${searchTerm}'`);
+
+        articlesToBeRestored.forEach((article) => {
+            hiddenArticles.splice(hiddenArticles.indexOf(article), 1);
```
Comment (Drag13) [inline src/sanitizer.js]: This code uses global variable `hiddenArticles` which complicates code maintaining. Using non-read-only global variables should be limited to the absolutely necessary cases only

File: src/storage.js [hunk 4]
```diff
@@ -81,35 +62,39 @@ export class Storage {
...
+            if ( JSON.stringify(nv) != JSON.stringify(ov) )
+                handler(nv, ov);
+        });
     }
 
     /** Private
      * @return {SanitizerSettings} user settings
      */
-    async _loadSettings() {
+    async loadSettings() {
```
Comment (Drag13) [inline src/storage.js:100]: This method is not private any more. JsDoc comment is redundant.

File: src/storage.js [hunk 5]
```diff
@@ -81,35 +62,39 @@ export class Storage {
...
+     * Subscribe to settings change event, watched by settings key
+     * @param {string} key
      * @param {Function} handler
      */
-    onChange(handler) {
-        this._handler = handler;
+    onSettingsChange(key, handler) {
+        chrome.storage.onChanged.addListener((changes) => {
+            const nv = changes.settings.newValue[key], ov = changes.settings.oldValue[key];
+            if ( JSON.stringify(nv) != JSON.stringify(ov) )
```
Comment (Drag13) [inline src/storage.js]: Please always use curly. This makes refactoring more safe.
I will update eslint rules for that

**Comment by Drag13** [comment]
Hi @ximeric 

I reviewed the code, my comments are above.

**In general I like new functionality to instantly show/hide articles - it is nice. I also liked the way you changed subscription to the storage change event - the new approach looks more clear to me.**

But the way you are operating with hidden and banned articles using global variables seems too risky for me. The new code is more complex and harder too read. I would omit using global variables directly. May be it is sensible to abstract them somehow and hide. May be there is another approach. But right now the code brings more technical debt than value (except this is the last commit ever :) )

Regrading using comma separated declaration block. I also love Pascal but for me, this approach works only in simple cases and simple methods. For the complex methods, when you are using `if` statement, or need to initialize variable using some condition it works not so good. One more point is that this requires to apply this approach to all code base and I don't see any reasonable reason for this extra work.

I would also kindly ask you to use prettier before pushing PR. It should work automatically with `pre-push` hook but I see that for some reason something went wrong. You can do this manually using `npm run prettier`;

**Comment by ximeric** [comment]
Hello @Drag13.

> But the way you are operating with hidden and banned articles using global variables seems too risky for me. ... I would omit using global variables directly. May be it is sensible to abstract them somehow and hide. May be there is another approach.

1. Original source code of this extension uses global variables, so why this method can't be applied with new code?
1. "Global" variables actually not so global in web extension because of browser limits access with very limited scope (no access from other parts of extension or any other page script).
1. Code always executed in the single thread (no unexpected changes by other parts of extension).
1. Programmer sees all of code in the single source file (for simple check of incorrect use of global variables).
1. Constants use meaningful names (invalid use of an constant can be always visible).
1. Repeated assign to a constant is not possible in JS.

So, use of global variables not so risky as described in the comment above.

> But right now the code brings more technical debt than value

Proposed code does not requires an additional "keep in mind" for the above reasons.

> The new code is more complex and harder too read

Can you give an 2-3 examples of such code proposed in this PR (without eslint directions on const definition)?

**Comment by Drag13** [comment]
I am probably a bad explainer. 

Everything you mentioned is correct. But my biggest concern is that we have 3 different global objects that are read **and** modified anywhere from the code however, it seems that they should be in consistent. 

```javascript
hiddenArticles
visibleArticles
settings
```

Such approach leads to mistakes and errors, because you have to remember to always sync them by the caller. Of course you can use global variables. But when doing so you have to minimize number of them and make usage as clear as possible, ideally leave no way to use them incorrectly.

**Comment by ximeric** [comment]
> make usage as clear as possible, ideally leave no way to use them incorrectly

For this the programmer has a brain. I think it have been applied for understanding any other code in the commits. Is it really so hard? And any other interactions with global variables (including settings, default settings, selectors) must be inspected, isn't it?

**Comment by ximeric** [comment]
Hi @Drag13.

New global variables were made for quick access to articles on the page. This method is not suitable and strictly incompatible with future planned features, so I decided to rewrite some parts of the patch.

**Comment by Drag13** [comment]
Hey, @ximeric 

Thanks for the update and excuse me for the late response - was busy on the project and we have a snow here - just can't miss it :)
I'll review and come back shortly.

**Pull request merged**